docs/source/jit_language_reference_v2.rst

```diff
@@ -0,0 +1,251 @@
+them assume the default type.
+
+::
+
+    import torch
+
+    def f(a, b):
+        # type: (torch.Tensor, int) → torch.Tensor
+        return a+b
+
+    m = torch.jit.script(f)
+    print("TorchScript:", m(torch.ones([6]), 100))
+
+**Example 3**
+
+This example caused a compilation error because it type annotates ``self``. The solution is to remove the type annotation and let its type be automatically
+inferred.
+
+::
+
+    import torch
+
+    @torch.jit.script
+    class MyClass(object):
+    # ERROR: do not annotate the type of self
+    def __init__(self: MyClass, x: int):
+        self.x = x
+
+    def inc(self, val: int):
+        self.x += val
+
+Annotate variables and data attributes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+In general, types of data attributes (including class and instance data attributes) and local variables can be automatically inferred from assignment statements.
+Sometimes, however, if a variable or attribute is associated with values of different types (e.g., as ``None`` or ``TensorType``), then they may need to be explicitly
+type annotated as a *wider* type such as ``Optional[int]`` or ``Any``.
+
+Local variables
+"""""""""""""""
+Local variables can be annotated according to Python3 typing module annotation rule, i.e.,
+
+::
+
+    LocalVarAnnotation := Identifier [":" TSType] "=" Expr
+
+In general, types of local variables can be automatically inferred. In some cases, however, programmers may need to annotate a multi-type for local variables
+that may be associated with different concrete types. Typical multi-types include ``Optional[T]`` and ``Any``.
+
+**Example**
+
+::
+
+    import torch
+
+    def f(a, setVal: bool):
+        value: Optional[torch.Tensor] = None
+        if setVal:
+            value = a
+        return value
+
+    ones = torch.ones([6])
+    m = torch.jit.script(f)
+    print("TorchScript:", m(ones, True), m(ones, False))
+
+Instance data attributes
+""""""""""""""""""""""""
+Instance data attributes can be annotated according to Python3 typing module annotation rules. Instance data attributes can be annotated (optionally) as final
+via ``torch.jit.Final``.
+
+::
+
+    "class" ClassIdentifier ["(object)"] ":"
+        InstanceAttrIdentifier ":" TSType
+
+        "def __init__(self ("," ParameterAnnotation)* "):"
+            "self." InstanceAttrIdentifier "=" Expr
+            ...
+
+where ``InstanceAttrIdentifier`` is the name of an instance attribute. For ``ModuleType`` classes, instance attributes may be declared as *final* via
+``torch.jit.Final``:
+
+::
+
+    "class" ClassIdentifier "(torch.nn.Module):"
+    InstanceAttrIdentifier ":" ["torch.jit.Final("] TSType [")"]
+    ...
+
+where ``InstanceAttrIdentifier`` is the name of an instance attribute and ``torch.jit.Final`` indicates that the attribute cannot be re-assigned outside
+of ``__init__`` or overridden in subclasses.
+
+**Example**
+
+TODO: Add the example
```

Comment:
IIRC, using class attribute annotation to hint type of instance attribute isn't supported yet

Reply:
Do we just delete this section, then?

Reply:
Add example that shows this for modules.